topics/wallet-interaction-standard.md

```diff
@@ -0,0 +1,255 @@
+
+## Summary
+
+This specification describes a communication protocol between dapps (decentralized applications) and wallets. It defines messages that both sides should use to interact with each other and provides guidelines on how to process them.
+
+## Terminology
+
+* wallet: A service that manages a user's keys and can sign and perform canister calls on their behalf.
+* relying party: A service that wants to request calls on a specific canister.
+
+## Messages
+
+### `permission`
+
+The purpose of the `permission` messages is to establish a connection between a relying party and a wallet, grant the relying party access to public parts of the user's identity and define the scope of actions the relying part is allowed to perform.
+
+#### Types
+
+- `text`: A plain `string` value.
+- `blob`: A `string` value describing binary data encoded in base64.
+
+#### Request
+
+`version` (`text`): The version of the standard used. If the wallet does not support the version of the request, it must send the `"VERSION_NOT_SUPPORTED"` error in response.
+
+`appMetadata`: Information about the relying party.
+- `name` (`text`): The user-friendly name of the relying party. 
+- `icon` (`text`, optional): An optional URL pointing to an icon resource representing the relying party.
+
+`networks`: A list of networks on which the relying party plans to operate.
+- `chainId`(`text`): The chain id of the network as described in the [CAIP-2]() standard.
+- `name` (`text`, optional): An optional user-friendly name of the network.
+- `rpcUrl` (`text`, optional): An optional custom RPC URL associated with the network.
+
+`scopes`: A list of permission scopes the relying party requires. Possible values: 
+- `"canister_call"`
+
+`challenge` (`blob`): A challenge used for the wallet to sign in order to prove its access to the identity. The challenge should be an array of 32 cryptographically random bytes generated from a secure random source by the sender of the request.
+
+#### Response
+
+`version` (`text`): The version of the standard used. It must match the `version` from the request.
+
+`appMetadata`: Information about the wallet.
+- `name` (`text`): The user-friendly name of the wallet.
+- `icon` (`text`, optional): An optional URL poiting to an icon resource representing the wallet.
+
+`networks`: A list of networks on which the user has agreed the relying party can operate. This should be a subset of the `networks` from the original request.
+- `chainId` (`text`): The chain id of the network as described in the [CAIP-2]() standard.
+- `name` (`text`, optional): An optional user-friendly name of the network.
+- `rpcUrl` (`text`, optional): An optional custom RPC URL associated with the network.
+
+`scopes`: A list of permission scopes that the user has agreed the relying party can be granted. This should be a subset of the `scopes` field from the original request. Possible values:
+- `"canister_call"`
+
+`identity`: The identity the user has selected.
```

Comment:
I think there is a use-case for sharing multiple keys at once. Imagine having a dapp that specializes in asset management. It would be nice if you could easily plug in your wallet that controls multiple keys into that dapp to control all your assets (spread across many keys) with a single connection request.

So my suggestion is to change `identity` to `identities` (array) containing objects of the same shape as the current `identity`.

Reply:
Yeah this also makes sense if the wallet derives multiple private keys instead of subaccounts like Plug does. Do we want to have optional names for identities?

Reply:
Hm, I'm currently leaning towards having key names be a wallet internal concept. The wallet should of course show the key name on the transaction approval screen.

But I have no strong opinion on that. As an optional field, I think it's good to have. Then it is still up to the wallet whether or not to provide it (or make it a user choice). 👍